sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,1045 @@
+         * <ul>
+         *     <li>Invoked by the upstream when it errors, or</li>
+         *     <li>Invoked by the drain-loop when it detects a 'non-retriable or retry exhaust' error, or </li>
+         *     <li>Invoked by the drain-loop when it detects the receiver signaled terminal completion  and retry
+         *     is disabled (i.e., NULL_RETRY_POLICY is set).</li>
+         * </ul>
+         *
+         * @param e the error signaled.
+         */
+        @Override
+        public void onError(Throwable e) {
+            if (done) {
+                Operators.onErrorDropped(e, messageSubscriber.currentContext());
+                return;
+            }
+
+            // It is possible that -
+            // 1. the error from upstream and 'non-retriable or retry exhaust' error from RetryLoop
+            // 2. Or, the error from upstream and error from receiver when NULL_RETRY_POLICY is set
+            // signals concurrently,  if so, the ERROR will be a CompositeException storing both errors.
+            if (Exceptions.addThrowable(ERROR, this, e)) {
+                done = true;
+                final String logMessage;
+                if (retryPolicy == NULL_RETRY_POLICY) {
+                    logMessage = "Terminal error signal from Upstream|Receiver arrived at MessageFlux.";
+                } else {
+                    logMessage = "Terminal error signal from Upstream|RetryLoop arrived at MessageFlux.";
+                }
+                mediatorHolder.withReceiverInfo(logger.atWarning()).log(logMessage, e);
+                drain(null);
+            } else {
+                // Once the drain-loop processed the last error, then further errors dropped through the standard
+                // Reactor channel. E.g., retry exhaust error happened and, as part of its processing, upstream
+                // gets canceled, but if upstream still signals an error, then that error gets dropped.
+                Operators.onErrorDropped(e, messageSubscriber.currentContext());
+            }
+        }
+
+        /**
+         * Signals operator termination with completion.
+         * <ul>
+         *     <li>Invoked by the upstream when it completes, or</li>
+         *     <li>Invoked by the the drain-loop when it detects first receiver signaled terminal completion and retry
+         *     is disabled (i.e., NULL_RETRY_POLICY is set).</li>
+         * </ul>
+         */
+        @Override
+        public void onComplete() {
+            if (done) {
+                return;
+            }
+
+            done = true;
+            final String logMessage;
+            if (retryPolicy == NULL_RETRY_POLICY) {
+                logMessage = "Terminal completion signal from Upstream|Receiver arrived at MessageFlux.";
+            } else {
+                logMessage = "Terminal completion signal from Upstream arrived at MessageFlux.";
+            }
+            mediatorHolder.withReceiverInfo(logger.atWarning()).log(logMessage);
```

Comment:
This is a terminal path, so I was wondering if this should be Warn or Info? For example, if there is an incident of receiver hang but also see this log, then it could be that downstream (e.g., user handler) of drain-loop is stuck. E.g., Useful in case of (non-rolling) session receiver where an error or completion of endpoint requires eventual stoppage of receiving. Let me know what you think, so we can adjust the level if needed.